cranelift/isle/isle/src/sema.rs

```diff
@@ -622,6 +674,55 @@ impl TypeEnv {
             }
         }
 
+        // Finally, collect converters.
+        for def in &defs.defs {
+            match def {
+                &ast::Def::Converter(ast::Converter {
+                    ref func,
+                    ref from_ty,
+                    ref to_ty,
+                    infallible,
+                    borrow,
+                    pos,
+                }) => {
+                    let from_ty = tyenv.intern_mut(from_ty);
+                    let from_ty = match tyenv.type_map.get(&from_ty) {
+                        Some(ty) => *ty,
+                        None => {
+                            tyenv.report_error(pos, "Unknown from-type for converter");
+                            continue;
+                        }
+                    };
+
+                    let to_ty = tyenv.intern_mut(to_ty);
+                    let to_ty = match tyenv.type_map.get(&to_ty) {
+                        Some(ty) => *ty,
+                        None => {
+                            tyenv.report_error(pos, "Unknown to-type for converter");
+                            continue;
+                        }
+                    };
+
+                    let func = tyenv.intern_mut(func);
+
+                    if tyenv.converters.contains_key(&(from_ty, to_ty)) {
+                        tyenv.report_error(pos, "Converter already exists for this type pair");
+                        continue;
+                    }
+
+                    tyenv.converters.insert(
+                        (from_ty, to_ty),
+                        Converter {
+                            func,
+                            infallible,
+                            borrow,
+                        },
+                    );
```

Comment:
This can avoid a double lookup by doing something like

```rust
let old_entry = tyenv.converters.insert(...);
if let Some(old_entry) = old_entry {
    tyenv.report_error(...);
    continue;
}
```

Reply:
Thanks! Used the entry API actually for a similarly cleaner single-lookup.